feat: Add 'root' and 'folder' filters

ilandikov opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

<!--- Describe your changes in detail -->

## Motivation and Context

Refactoring `Group` methods to `Field` methods need creation of new fields
Part of #920, continuing #1870.

I took the base class for the new Field as `TextField` since it's `grouper()` was already implemented  there

The contents of `groupBy...()` methods were moved to `value()` of the new classed, obviously removing the array in the `return` (It will be added in `TextField.grouper()`). This is a major difference with the #1870 where the methods have been moved to `grouper()` of the existing Field classes.

As for the tests, I reused the tests from `Group.test` and also created new tests to test the `value()` in a simple manner since this is where the change is done.

I see more refactoring to be done in `Group` but I think this PR is already big enough so I'll save that for a later one.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [ ] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [x] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [ ] **Documentation** (prefix: `docs` - improvements to any documentation content **for users**)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))
- [ ] **Contributing Guidelines** (prefix: `contrib` - any improvements to documentation content **for contributors** - see [Contributing to Tasks](https://publish.obsidian.md/tasks-contributing/))

Internal changes:

- [x] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [x] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [x] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://publish.obsidian.md/tasks-contributing/Documentation/About+Documentation) accordingly.
- [x] My change has adequate [Unit Test coverage](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://publish.obsidian.md/tasks-contributing) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://publish.obsidian.md/tasks-contributing)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

claremacrae reviewed (changes requested):
Nice.

By adding a new implementation of `TextField` and adding it to `FilterParser`, you have added support for filtering and sorting of 3 new fields! ❤️ 🎉 

So that means there is more work to be done ... In the interested of keeping `main` releasable, I would like it to be done in this PR please....

Take a deep breath.... Here are the instructions:

https://publish.obsidian.md/tasks-contributing/Code/How+do+I+add+a+new+task+filter

(I just fixed some broken links in that page, so use the online version, not the not the one in your fork)

If you have the time, could you update the summary line of this PR too...

ilandikov commented:
> If you have the time, could you update the summary line of this PR too...

Oops, I got too excited and forgot this one =)

> Take a deep breath.... Here are the instructions:

Love it =)

> you have added support for filtering and sorting of 3 new fields

Oh wow...This was..._simple_ ;)

Ok so I'm missing things tests and docs on filtering and sorting of the new fields?

claremacrae commented:
Hi @ilandikov I think I'm wrong about this having enabled sorting...

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/cfa3f8884be7f19d1df906b68159a7054d28f4ac/src/Query/Filter/TextField.ts#L95-L98

But I think it probably has enabled filtering... Maybe????

ilandikov commented:
Well, it’s a matter of enabling sorting (new information) and writing tests (as discussed already)

claremacrae commented:
Sorry to be the bearer of bad news....

There were no tests for filtering in Root and Folder, so I added tests.

I was unable to get the Root ones to pass, and after some time, added more tests its value() which showed that there is an error in the `RootField.value()`

When fix it, please revert the duplication of `value()` implementation in this class. The grouping code can simply call `this.value()` in this simple case.

claremacrae commented on src/Query/Filter/BacklinkField.ts:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public grouper(): GrouperFunction {
+        return (task: Task) => {
+            const linkText = task.getLinkText({ isFilenameUnique: true });
+            if (linkText === null) {
+                return ['Unknown Location'];
+            }
+
+            let filenameComponent = 'Unknown Location';
+
+            if (task.filename !== null) {
+                // Markdown characters in the file name must be escaped.
+                filenameComponent = Group.escapeMarkdownCharacters(task.filename);
+            }
+
+            if (task.precedingHeader === null || task.precedingHeader.length === 0) {
+                return [filenameComponent];
+            }
+
+            // Markdown characters in the heading must NOT be escaped.
+            const headingGrouper = new HeadingField().createGrouper().grouper;
+            const headingComponent = headingGrouper(task)[0];
+
+            if (filenameComponent === headingComponent) {
+                return [filenameComponent];
+            } else {
+                return [`${filenameComponent} > ${headingComponent}`];
+            }
```
I am uncomfortable at the amount of duplicated code between value() and grouper() here.

claremacrae commented on src/Query/Filter/FolderField.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public grouper(): GrouperFunction {
+        return (task: Task) => {
+            const path = task.path;
+            const fileNameWithExtension = task.filename + '.md';
+            const folder = path.substring(0, path.lastIndexOf(fileNameWithExtension));
+            if (folder === '') {
+                return ['/'];
+            }
+            return [Group.escapeMarkdownCharacters(folder)];
+        };
+    }
+}
```
I am uncomfortable with the amount of duplicated code between value() and grouper().

claremacrae commented on src/Query/Filter/RootField.ts:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public grouper(): GrouperFunction {
+        return (task: Task) => {
+            const path = task.path.replace(/\\/g, '/');
+            const separatorIndex = path.indexOf('/');
+            if (separatorIndex == -1) {
+                return ['/'];
+            }
+            return [Group.escapeMarkdownCharacters(path.substring(0, separatorIndex + 1))];
+        };
```
Again, I am uncomfortable at the amount of duplicated code here.

claremacrae commented on tests/Query/Filter/FolderField.test.ts:
```diff
@@ -0,0 +1,55 @@
+import { fromLine } from '../../TestHelpers';
+
+describe('folder', () => {
+    it('should provide access to the file name with extension', () => {
```
Remove `file name` reference

ilandikov commented:
> Sorry to be the bearer of bad news....
> 
> There were no tests for filtering in Root and Folder, so I added tests.
> 

That's actually good news =) Bug report on a released feature would be bad news ;)

> I was unable to get the Root ones to pass, and after some time, added more tests its value() which showed that there is an error in the `RootField.value()`

Fixed. Btw are you sure Obsidian would provide roots starting with `/`.

> When fix it, please revert the duplication of `value()` implementation in this class. The grouping code can simply call `this.value()` in this simple case.

Indeed. I did this for the `FolderField` as well.
